File: python/atelier_bench/metrics/payment_delay.py
"""payment-delay metric."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.metric_contract import MetricUnit
from ..support.abstract_metric import AbstractMetric

if TYPE_CHECKING:
    from ..dataset import Dataset


class PaymentDelayMetric(AbstractMetric):
    KEY = "payment-delay"

    def __init__(self) -> None:
        super().__init__(self.KEY, MetricUnit.DAYS)

    def compute(self, data: "Dataset") -> float:
        return sum(1 for invoice in data.invoices if not invoice.paid) * 3.5

File: go/atelier/metrics/payment_delay.go
// Package metrics carries the 16 metrics. Each embeds support.BaseMetric,
// declares only Compute, and registers itself in an init function.
package metrics

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction; nothing else names the contract here.
var _ contracts.MetricContract = (*PaymentDelayMetric)(nil)

// PaymentDelayMetric computes the payment-delay metric.
type PaymentDelayMetric struct {
	support.BaseMetric
}

// NewPaymentDelayMetric builds the metric with its key and unit.
func NewPaymentDelayMetric() *PaymentDelayMetric {
	return &PaymentDelayMetric{BaseMetric: support.NewBaseMetric("payment-delay", contracts.UnitDays)}
}

// Compute is the only contract method this type declares; Key and Unit are
// promoted from the embedded base.
func (m *PaymentDelayMetric) Compute(data *dataset.Dataset) float64 {
	unpaid := 0
	for _, invoice := range data.Invoices {
		if !invoice.Paid {
			unpaid++
		}
	}
	return float64(unpaid) * 3.5
}

// init registers this metric.
func init() {
	Register(NewPaymentDelayMetric())
}
